### ISSUE TITLE
Add custom emojis support to the display of user names

### ISSUE THREAD

**Issue by SpcCw** [friendica/friendica#11991]
Some platforms allow to use custom emojis (specific for their server) and more and more users start to add them to their names. Since Friendica can't display these emojis it shows their codes instead sometimes resulting in super long and weirdly looking user names.

This can break page layout and make some threads hard to read, especially on mobile. Especially with user name badges which do not warp. Some deeper threads can appear completely broken because of this.

Examples:
![image](https://user-images.githubusercontent.com/35988439/195782450-b06298a7-e023-4c4e-b42d-57f3a1287f49.png)
![image](https://user-images.githubusercontent.com/35988439/195782623-f1f61b3e-33ca-4c41-8b65-ddb6a7cd0161.png)

**Possible solutions used by other platforms:**

- To trim these codes from user names if the server can't display these emojis so only the normal text is shown
- To shorten long usernames to certain length displaying them like "superlonguserna..."
- Maybe both of the above as text only super long usernames also happen and they can break page layout as well

**Comment by MrPetovan** [comment]
I started looking into importing custom emojis from remote servers to be able to convert these to images but I stopped in-between. There is a dilemma we need to solve first:
- Either we store the emoji images locally and it could grow extremely large because it depends on remote servers.
- Or we link to the the remote server asset and we have a privacy issue on our hands.

As for your suggestions:
- We can't guess what codes are emojis or plain text, the format is not specific enough across servers.
- Why not display ellipses, but coming up with a good maximum width in all contexts is a doozy

**Comment by annando** [comment]
There is the Mastodon API endpoint `api/v1/custom_emojis`. We regularly poll the servers anyway. So we could store the used shortcode anyway. During the contact discovery we then could simply remove these shortcodes.

**Comment by MrPetovan** [comment]
This doesn't resolve the dilemma: store assets locally or link to remote domains?

**Comment by SpcCw** [comment]
>  store assets locally or link to remote domains?

I would expect it to follow avatar cache logic (and even use the same setting): if caching is on then it is better to store them because other servers can be wonky at times and it can affect both experience and performance, if caching is off then same thing as with avatars.

Avatar cache already gets so large I doubt adding emojis will make much difference. But it is definitely useful.

**Comment by annando** [comment]
We will not display these graphics at all, so there is no need to store them locally.

**Comment by MrPetovan** [comment]
Why not? If we hit the custom emojis endpoint and store the shortcodes, we're halfway there! We already display these graphics in posts, why not in usernames?

**Comment by annando** [comment]
These "emojis" are not real emojis but graphics:
![grafik](https://user-images.githubusercontent.com/844208/195988150-7ee71a23-a618-49ce-a680-05425ecc4918.png)

The contact names are stored as raw text and isn't interpreted as HTML.

**Comment by MrPetovan** [comment]
I know but it's arbitrary, we can perfectly add the emojis images if we want to.

**Comment by ne20002** [comment]
I agree. I'm using (mostly) Tusky with Friendica. Custom emojis from Mastodon accounts are not displayed correctly / as expected in the app and it is a bit annoying as these are sometimes relevant for context (I see them as text with colons in account names and as blue boxes in posts).
As these are pictures I would treat them as exactly this and download and cache them as it is done with pictures in posts.

**Comment by eengnr** [comment]
I have a few contacts which show their antipathy against certain political parties with custom emojis in the name. But as Friendica does not load the emojis and instead only shows the shortcode, the opposite effect takes place - every time I see "advertisement" for the political parties. And this is of course not the intention behind the specific custom emoji.

It would be nice, if this could be fixed somehow.

**Comment by MrPetovan** [comment]
Here's the tentative synopsis for this task:
- Create a new database table `custom_emojis` with the following fields:
    - `id`
    - `gsid` referencing `gserver.id`
    - `shortcode`
    - `original_url`
    - `static_url`
    - `timestamp`
- During server update:
    - Call `/api/v1/custom_emojis`
    - If successful, remove all rows in `custom_emojis` for the relevant `gsid` if it isn't the local node and store the data in the `custom_emojis` table.
- On contact display name generation:
    - HTML escape the display name
    - **then** perform shortcode replace with `<img>` based on the contact's `gsid`
    - Remove the automatic escaping done in templates for the contact display name

Performance considerations: The contact name replacement function needs to cache the shortcode list for any given `gsid`. Or we could store the converted display name in the contact table, but this would require to regenerate all the contact display names for a given server every time we are updating the server.

**Comment by MrPetovan** [comment]
I just learned [`pixelfed.social` supports 47,700+ custom emojis](https://mastodon.social/@cheeaun/112291603392332206) which makes the just-in-time method replacement of emojis in display names a performance hazard. 😰

**Comment by csolisr** [comment]
> I just learned [`pixelfed.social` supports 47,700+ custom emojis](https://mastodon.social/@cheeaun/112291603392332206) which makes the just-in-time method replacement of emojis in display names a performance hazard. 😰

So plan B is to perform the shortcode HTML replacement directly into the `contacts` table, on the contact update step?

**Comment by MrPetovan** [comment]
Indeed, keeping a record of the original display name and the HTML version once the custom emojis have been replaced.
